src/connection.ts

```diff
@@ -1,71 +1,197 @@
-	connection.onShutdown(() => handler.shutdown());
-
-	// textDocument
-	connection.onDidOpenTextDocument(params => handler.didOpen(params));
-	connection.onDidChangeTextDocument(params => handler.didChange(params));
-	connection.onDidSaveTextDocument(params => handler.didSave(params));
-	connection.onDidCloseTextDocument(params => handler.didClose(params));
-	connection.onReferences((params, token) => handler.getReferences(params, token));
-	connection.onHover((params, token) => handler.getHover(params, token));
-	connection.onDefinition((params, token) => handler.getDefinition(params, token));
-	connection.onDocumentSymbol((params, token) => handler.getDocumentSymbol(params, token));
-	connection.onCompletion((params, token) => handler.getCompletions(params, token));
-	connection.onRequest('textDocument/xdefinition', (params, token) => handler.getXdefinition(params, token));
-
-	// workspace
-	connection.onWorkspaceSymbol((params, token) => handler.getWorkspaceSymbols(params, token));
-	connection.onRequest('workspace/xreferences', (params, token) => handler.getWorkspaceReference(params, token));
-	connection.onRequest('workspace/xpackages', (params, token) => handler.getPackages(params, token));
-	connection.onRequest('workspace/xdependencies', (params, token) => handler.getDependencies(params, token));
+export function registerLanguageHandler(
+	messageEmitter: MessageEmitter,
+	messageWriter: MessageWriter,
+	handler: TypeScriptService,
+	logger: Logger = new NoopLogger(),
+	tracer = new Tracer()
+): void {
+	/** Tracks Subscriptions for results to unsubscribe them on $/cancelRequest */
+	const subscriptions = new Map<string | number, Subscription>();
+	messageEmitter.on('message', async message => {
+		logger.log('-->', message);
+		// Ignore responses
+		if (!isRequestMessage(message) && !isNotificationMessage(message)) {
+			return;
+		}
+		if (message.method === '$/cancelRequest' && isNotificationMessage(message)) {
+			// Cancel another request by unsubscribing from the Observable
+			const subscription = subscriptions.get(message.params.id);
+			if (!subscription) {
+				logger.error(`$/cancelRequest for unknown request ID ${message.params.id}`);
+				return;
+			}
+			subscription.unsubscribe();
+			subscriptions.delete(message.params.id);
+			messageWriter.write({
+				jsonrpc: '2.0',
+				id: message.params.id,
+				error: {
+					message: 'Request cancelled',
+					code: ErrorCodes.RequestCancelled
+				}
+			} as ResponseMessage);
+			return;
+		}
+		// If message is request and has tracing metadata, extract the span context and create a span for the method call
+		let span = new Span();
+		if (hasMeta(message) && isRequestMessage(message)) {
+			const context = tracer.extract(FORMAT_TEXT_MAP, message.meta);
+			if (context) {
+				span = tracer.startSpan('Handle ' + message.method, { childOf: context });
+				span.setTag('params', inspect(message.params));
```

Comment:
Yeah, but it's intentional here. The tag values are shown in a single line, so limiting the depth makes sense.